### PR TITLE
Add etcdv3 client

### PR DESCRIPTION
**Pull request by ajroetker**
This PR updates the etcd metafora client to use the v3 apis and client.
The main notes here are that we can avoid certain race conditions now with transactions and the watcher code is tremendously simpler.

### CODE REVIEW

File: metcdv3/balancer.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	// Get the list of all claimed work, create a map of the counts and
+	// node values
+	// We ignore tasks which have no claims
+	for _, kv := range resp.Kvs {
+		ownerPath := path.Base(string(kv.Key))
+		if ownerPath == OwnerPath {
+			ov := &ownerValue{}
+			err := json.Unmarshal(kv.Value, &ov)
```
Comment (mattayes) [inline metcdv3/balancer.go]: [NIT] `ov` is already a pointer; do you need another `&`?

```go
err := json.Unmarshal(kv.Value, ov)
```

File: metcdv3/client.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	"path"
+	"strconv"
+	"time"
+
+	etcdv3 "github.com/coreos/etcd/clientv3"
+	"github.com/lytics/metafora"
+)
+
+var (
+	//ErrNotOwner                    = errors.New("metafora etcd clientv3: not owner")
```
Comment (mattayes) [inline metcdv3/client.go]: Do we not use these?

File: metcdv3/client.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+// SubmitTask creates a new task in etcd
+func (mc *mclient) SubmitTask(task metafora.Task) error {
+	c := context.Background()
+	fullpath := path.Join(mc.taskPath(task.ID()), PropsPath)
+	buf, err := json.Marshal(task)
+	if err != nil {
+		return err
+	}
+	txnRes, err := mc.kvc.Txn(c).
+		If(etcdv3.Compare(etcdv3.Version(fullpath), "=", 0)).
```
Comment (mattayes) [inline metcdv3/client.go]: Whoa, I didn't know you could do this in Go! 🤯

**Comment by ajroetker** [comment]
@mattayes Thanks for the feedback, addressed your comments!

File: metcdv3/balancer.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		return nil, err
+	}
+
+	if resp == nil || len(resp.Kvs) == 0 {
+		metafora.Warnf("balancer received empty response from GET %s", e.nodePath)
+		return state, nil
+	}
+
+	for _, kv := range resp.Kvs {
+		// We're garunteed to find nodes under the _metadata path (created on Coordinator startup)
```
Comment (timkaye11) [inline metcdv3/balancer.go]: typo :wink:

Comment (ajroetker) [reply metcdv3/balancer.go]: The only thing we we're missing was u ;)

File: metcdv3/coordinator.go [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+
+	parseTask := func(we *WatchEvent) metafora.Task {
+		// Pickup new tasks
+		if we.Type != Delete {
+			dir, taskID := path.Split(we.Key)
+			dir = path.Clean(dir)
+			if taskID == PropsPath {
+				_, taskID := path.Split(dir)
+				metafora.Debugf("%s received task with properties: %s", ec.name, taskID)
+				fmt.Printf("%s received task with properties: %s\n", ec.name, taskID)
```
Comment (kyledj) [inline metcdv3/coordinator.go]: Printf + Debugf?

File: metcdv3/coordinator.go [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+				if resp, err := ec.kvc.Get(c, path.Join(dir, OwnerPath), etcdv3.WithLimit(1)); err != nil {
+					metafora.Errorf("%s error getting properties while handling %s", ec.name, taskID)
+					fmt.Printf("%s error getting properties while handling %s\n", ec.name, taskID)
```
Comment (kyledj) [inline metcdv3/coordinator.go]: Same

File: metcdv3/coordinator.go [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+					return nil
+				} else if resp.Count == 1 {
+					metafora.Debugf("%s received claimed task: %s", ec.name, taskID)
+					fmt.Printf("%s received claimed task: %s\n", ec.name, taskID)
```
Comment (kyledj) [inline metcdv3/coordinator.go]: Same

**Comment by mattayes** [comment]
:point_down: Looks like there's a race condition :grimacing:

**Pull request merged**